src/simpleppl.jl

```diff
@@ -0,0 +1,749 @@
+# Adding interface for DAG representation using parts of SimplePPL/Mamba.jl
+using Graphs
+using Distributions
+
+
+ElementOrVector{T} = Union{T,Vector{T}}
+
+#################### Variate Types ####################
+
+abstract type ScalarVariate <: Real end
+abstract type ArrayVariate{N} <: DenseArray{Float64,N} end
+
+const AbstractVariate = Union{ScalarVariate,ArrayVariate}
+const VectorVariate = ArrayVariate{1}
+const MatrixVariate = ArrayVariate{2}
+
+
+#################### Distribution Types ####################
+
+const DistributionStruct =
+    Union{Distribution,Array{UnivariateDistribution},Array{MultivariateDistribution}}
+
+#################### Null Distribution ####################
+
+struct NullUnivariateDistribution <: UnivariateDistribution{ValueSupport} end
+
+
+#################### Dependent Types ####################
+
+mutable struct ScalarLogical <: ScalarVariate
+    value::Float64
+    symbol::Symbol
+    monitor::Vector{Int}
+    eval::Function
+    sources::Vector{Symbol}
+    targets::Vector{Symbol}
+end
+
+mutable struct ArrayLogical{N} <: ArrayVariate{N}
+    value::Array{Float64,N}
+    symbol::Symbol
+    monitor::Vector{Int}
+    eval::Function
+    sources::Vector{Symbol}
+    targets::Vector{Symbol}
+end
+
+mutable struct ScalarStochastic <: ScalarVariate
+    value::Float64
+    symbol::Symbol
+    monitor::Vector{Int}
+    eval::Function
+    sources::Vector{Symbol}
+    targets::Vector{Symbol}
+    distr::UnivariateDistribution
+end
+
+mutable struct ArrayStochastic{N} <: ArrayVariate{N}
+    value::Array{Float64,N}
+    symbol::Symbol
+    monitor::Vector{Int}
+    eval::Function
+    sources::Vector{Symbol}
+    targets::Vector{Symbol}
+    distr::DistributionStruct
+end
+
+const AbstractLogical = Union{ScalarLogical,ArrayLogical}
+const AbstractStochastic = Union{ScalarStochastic,ArrayStochastic}
+const AbstractDependent = Union{AbstractLogical,AbstractStochastic}
+
+#################### Variate ####################
+
+#################### Conversions ####################
+
+Base.convert(::Type{Bool}, v::ScalarVariate) = convert(Bool, v.value)
+Base.convert(::Type{T}, v::ScalarVariate) where {T<:Integer} = convert(T, v.value)
+Base.convert(::Type{T}, v::ScalarVariate) where {T<:AbstractFloat} = convert(T, v.value)
+Base.AbstractFloat(v::ScalarVariate) = convert(Float64, v)
+Base.Float64(v::ScalarVariate) = convert(Float64, v)
+
+Base.convert(::Type{Matrix}, v::MatrixVariate) = v.value
+Base.convert(::Type{Vector}, v::VectorVariate) = v.value
+Base.convert(
+    ::Union{Type{Array{T}},Type{Array{T,N}}},
+    v::ArrayVariate{N},
+) where {T<:Real,N} = convert(Array{T,N}, v.value)
+
+Base.unsafe_convert(::Type{Ptr{Float64}}, v::ArrayVariate) = pointer(v.value)
+
+
+macro promote_scalarvariate(V)
+    quote
+        Base.promote_rule(::Type{$(esc(V))}, ::Type{T}) where {T<:Real} = Float64
```

Comment:
This seems wrong in general - I don't think dual numbers or eg `BigFloat` should be promoted to `Float64`.